java/src/jmri/jmrit/operations/rollingstock/RollingStock.java

```diff
@@ -813,7 +815,7 @@ public void propertyChange(java.beans.PropertyChangeEvent e) {
                         if (e.getNewValue() != null) {
                             Location newLocation = locationManager.getLocationByReporter((jmri.Reporter) e.getNewValue());
                             if (newLocation != getLocation())
-                                setLocation(newLocation, null);
+                                setWhereLastSeen(newLocation);
```

Comment:
On 08/10/2016 08:23 PM, Daniel Boudreau wrote:
> From: Paul Bender [mailto:notifications@github.com]
>> That's actually overkill. You can build a wheel report by track with a
>> single reader at the yard thought using switch position information to
>> determine
> That’s great way of doing it, is there something already coded
> that maps switch positions to a track? I want to pass this along to him.

The idea is that you create an internal reader that is associated with 
each of the staging tracks.  You use a script to receive the data from 
the real reader and pass that along to the correct internal reader.  I 
probably have time to mach that one up over the next day or two.

(I was going to suggest using a Logix for this, but Logix won't actually 
do anything with readers right now, and I'm not going to suggest making 
any changes to logix until we have a testing infrastructure in place 
that can make heads or tails of the existing classes... )

Now, there is one missing detail in all this.  We're currently 
associating tag readers with locations.  This choice was made because a 
point tag reader (such as RFID) can be installed on the mainline at a 
location (which isn't a track used for operations) and area tag readers 
( such as Digitrax's transponding or Lenz's Railcom ) can't tell what 
track within a zone something is on (if there is more than one track in 
the zone).

If the user really wants to have one reader per track (and here it 
doesn't matter if we are talking about real readers or virtual readers) 
we also need to be able to associate a reader with a track.

Seeing a car pass over a location associated reader should update the 
location of the car.  Seeing a car pass over a track associated reader 
should update both the location and the track the car is on.  The code 
(both before you changed it and now) can't actually handle updating the 
track the car is on.

 > Do you always create a wheel report immediately after a train passes 
a reader?

Typically, yes.

> I see a race condition between two update mechanisms, does that make sense to you?

Ok. I see that now.

> I agree
> that I should break existing code, but I’m thinking that it’s broke the
> way it was written and I’m trying to fix it.Given that entering a car’s
 > tag Id from operations has been broken  > for a while, I don’t think 
we have a lot of users using tags and operations.

I'm not actually sure this has been broken for long.  I didn't have a 
chance to look at this before you made changes to fix the code, but I 
know it was working the last time I was using it (sometime before early 
June, when I started getting modules ready for the NTS).

> Not sure when and how
> your script creates the wheel report, but if it doesn’t react
> immediately to a car location change, there’s a chance that the train
> move update will overwrite the location and date.

I see what you're saying, and I can update my scripts to use I'll update 
my scripts to use the getWhereLastSeen and getWhenLastSeen methods, but 
that doesn't actually solve the problem.  I think those methods should 
be re-written as wrappers that pull data from the IDTags, and the 
setWhereLastSeen and setWhenLastSeen methods should be removed, in favor 
of letting the readers updates through the event notification.

At a minimum, the operations code doesn't start until you actually open 
one of the operations tables, so if the operations code is started after 
you've already started moving trains on the layout, the values you're 
storing in the operations files are no longer correct.

In other words, the ID Tag should be the entity that we trust most with 
the when and where something was last seen, not the piece of rolling stock.

If the user needs this information, he can get the information from the 
IDTag XML file (as long as he remembers to set the checkbox to store the 
when/where last seen information).

(This also goes to Bob's point about separating the automated update 
process from the perception generated by the operations code.)

In order to handle finding the track the car was last seen on,  which I 
noted above, we also need to add a getTrackWhereLastSeen method, which 
would also be a wrapper that pulls information from the IDTag data.

Paul